gateway-provider-security-jwt/src/main/java/org/apache/knox/gateway/provider/federation/jwt/filter/AbstractJWTFilter.java

```diff
@@ -407,7 +415,7 @@ protected boolean verifyTokenSignature(final JWT token) {
     String tokenId = TokenUtils.getTokenId(token);
 
     // Check if the token has already been verified
-    verified = hasSignatureBeenVerified(tokenId);
+    verified = (tokenId != null) && hasSignatureBeenVerified(tokenId);
```

Comment:
if we do decide to support 3rd party tokens, we would likely want to make sure that such tokens also benefit from the caching. Which seems to imply that we would either need to derive a tokenId from the 3rd party token somehow or use the actual token as the id. The latter would have uniqueness implications across jobs but may be okay for this optimization. If the token is exactly the same then it is the same user and created at exactly the same time with the exact same expiry?

Reply:
This is a valid point, and one that I would argue can be addressed as part of revisiting the the signature verification caching implementation altogether.

Reply:
I can buy that.